src/pages/ReportsUpload/index.tsx

```diff
@@ -0,0 +1,341 @@
+import { GlobalState } from '../../models/state';
+import { RouterGlobalProps } from '../../models/router';
+import {
+    Main,
+    PageHeader,
+    PageHeaderTitle
+} from '@redhat-cloud-services/frontend-components';
+import { Upload } from '../../models';
+import './ReportsUpload.scss';
+import { UploadForm } from './UploadForm';
+
+interface FormValues {
+    file: string;
+    reportName: string;
+    reportDescription: string;
+    yearOverYearGrowthRatePercentage: number;
+    percentageOfHypervisorsMigratedOnYear1: number;
+    percentageOfHypervisorsMigratedOnYear2: number;
+    percentageOfHypervisorsMigratedOnYear3: number;
+    percentageOfHypervisorsMigratedSum: number // Sum of year1, year2, and year3
+}
+
+interface StateToProps {
+    file: File;
+    success: boolean | null;
+    error: string | null;
+    progress: number;
+    uploading: boolean
+}
+
+interface DispatchToProps {
+    uploadProgress: (progress: number) => void;
+    uploadRequest: (upload: Upload, config: {}) => void;
+    selectUploadFile: (file: File) => void;
+}
+
+interface Props extends StateToProps, DispatchToProps, RouterGlobalProps {
+}
+
+interface State {
+    showForm: boolean;
+    timeoutToRedirect: number;
+}
+
+export class ReportsUpload extends React.Component<Props, State> {
+
+    redirectTimer: any;
+    beforeUnloadHandler: any;
+
+    initialFormValue: FormValues;
+
+    constructor(props: Props) {
+        super(props);
+
+        this.state = {
+            showForm: true,
+            timeoutToRedirect: 3
+        };
+
+        this.initialFormValue = {
+            file: '',
+            reportName: '',
+            reportDescription: '',
+            yearOverYearGrowthRatePercentage: 5,
+            percentageOfHypervisorsMigratedOnYear1: 50,
+            percentageOfHypervisorsMigratedOnYear2: 30,
+            percentageOfHypervisorsMigratedOnYear3: 10,
+            percentageOfHypervisorsMigratedSum: 90
+        };
+
+        this.redirectTimer = null;
+    }
+
+    componentDidMount() {
+        this.beforeUnloadHandler = window.addEventListener('beforeunload', (event) => {
+            if (this.props.uploading) {
+                event.preventDefault();
+                return event.returnValue = 'Are you sure you want to close? The upload has not finished yet';
+            }
+
+            return event;
+        });
+    }
+
+    componentWillUnmount() {
+        removeEventListener('beforeunload', this.beforeUnloadHandler);
+    }
+
+    /**
+     * Called by "Close modal" or by the "Cancel button"
+     */
+    handleCloseModel = () => {
+        if (!this.props.uploading) {
+            this.props.history.push('/reports');
+        }
+    };
+
+    actionsOnUploadSuccess = () => {
+        const { timeoutToRedirect } = this.state;
+
+        this.startRedirectTimer();
+
+        if (this.state.timeoutToRedirect === 0) {
+            this.props.history.push('/reports');
```

Comment:
Hmm. I think it's a React anti-pattern to have code with side-effects called in the render method (which it effectively is here since `actionsOnUploadSuccess` is called in `render()`.

https://reactjs.org/docs/react-component.html#render

> The render() function should be pure, meaning that it does not modify component state, it returns the same result each time it’s invoked, and it does not directly interact with the browser.

What you're doing here is checking whether it's time to redirect yet whenever the button renders, but what you really want to do is check and do the redirect when the component updates (when the state changes). So you probably want to add a `componentDidUpdate(prevProps, prevState)` lifecycle method and do your redirect in there, behind a condition like `if(this.props.success && prevState.timeoutToRedirect !== 0 && this.state.timeoutToRedirect === 0)`.

The reason to do this in componentDidUpdate, and to check `prevState` to make sure it's the latest update that caused the timer to hit zero, is to prevent unintentionally doing multiple redirects or possibly getting yourself into a render loop.

Reply:
This was a real mistake. You are totally right... I've just added `componentDidUpdate(prevProps, prevState)`